src/Raven.Server/Documents/Handlers/AI/Agents/AbstractAiAgentProcessor.cs

```diff
@@ -9,47 +9,163 @@
 using Raven.Server.Documents.AI;
 using Raven.Server.Documents.AI.AiGen;
 using Raven.Server.Documents.Handlers.Processors;
-using Raven.Server.ServerWide;
 using Raven.Server.ServerWide.Context;
 using Sparrow.Json;
 using Sparrow.Json.Parsing;
 using Sparrow.Server.Json.Sync;
 using System.Net;
 using Raven.Client.Exceptions;
+using Raven.Client.Exceptions.Documents;
+using Raven.Client.Json.Serialization;
 using Raven.Server.Documents.Handlers.Processors.MultiGet;
 using Sparrow;
 
 namespace Raven.Server.Documents.Handlers.AI.Agents
 {
     internal abstract class AbstractAiAgentProcessor : AbstractDatabaseHandlerProcessor<DatabaseRequestHandler, DocumentsOperationContext>
     {
+        protected AbstractAiAgentProcessor([NotNull] DatabaseRequestHandler requestHandler) : base(requestHandler)
+        {
+        }
 
-        public AbstractAiAgentProcessor([NotNull] DatabaseRequestHandler requestHandler) : base(requestHandler)
+        public async Task HandleRequest(
+            JsonOperationContext context, 
+            AiAgentConfiguration configuration, 
+            string chatId, 
+            ChatDocument chatDocument, 
+            RequestBody body,
+            CancellationToken token)
         {
+            var hasActionResponse = body.ActionResponses is { Length: > 0 };
+            var hasUserPrompt = string.IsNullOrEmpty(body.UserPrompt) == false;
+
+            if (hasActionResponse && hasUserPrompt)
+                throw new InvalidOperationException("Cannot have a chat with open tool calls and user prompt.");
+
+            if (body.ActionResponses != null)
+            {
+                foreach (BlittableJsonReaderObject tool in body.ActionResponses)
+                {
+                    var t = JsonDeserializationClient.ToolResponse(tool);
+                    if (chatDocument.OpenToolCalls.Remove(t.ToolId) == false)
+                        throw new InvalidOperationException($"{t.ToolId} is an unknown tool ID");
+
+                    chatDocument.Messages.Add(context.ReadObject(new DynamicJsonValue { ["tool_call_id"] = t.ToolId, ["role"] = "tool", ["content"] = t.Content },
```

Comment:
we shouldn't hardcode all over the place these names. We can create an instance of a class that will act as a DTO, call method .ToJson for example